ksqldb-engine/src/main/java/io/confluent/ksql/physical/scalablepush/ScalablePushRegistry.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.physical.scalablepush;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.GenericKey;
+import io.confluent.ksql.GenericRow;
+import io.confluent.ksql.execution.streams.materialization.Row;
+import io.confluent.ksql.execution.streams.materialization.TableRow;
+import io.confluent.ksql.execution.streams.materialization.WindowedRow;
+import io.confluent.ksql.physical.scalablepush.locator.AllHostsLocator;
+import io.confluent.ksql.physical.scalablepush.locator.PushLocator;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.util.PersistentQueryMetadata;
+import io.vertx.core.impl.ConcurrentHashSet;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.ProcessorSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This registry is kept with every persistent query, peeking at the stream which is the output
+ * of the topology. These rows are then fed to any registered ProcessingQueues where they are
+ * eventually passed on to scalable push queries.
+ */
+public class ScalablePushRegistry {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ScalablePushRegistry.class);
+
+  private final KStream<?, GenericRow> stream;
+  private final LogicalSchema logicalSchema;
+  private final PushLocator pushLocator;
+  private final Set<ProcessingQueue> processingQueues = new ConcurrentHashSet<>();
```

Comment:
it's a little odd for us to use a `Set` here when `PorcessingQueue` doesn't implement `hashCode` or `equals` since we'll never be able to check set membership 😉 that's my long way of saying, can this just be a list? (that also allows us to play around with guaranteeing iteration order)

Reply:
> Basically, I just used a Set so that I could have a constant time unregister rather than searching through a list.

I missed that - if that's the case then a reference set is a-ok